### PR TITLE
chore: refactor experiment creation modal and add continue workflow

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

as part of https://determinedai.atlassian.net/browse/DET-3703

this excludes the helper form that sets max steps and description

## Commentary (optional)

an important part of this is to create the new experiment config based on the trial and that happens here

```
          const newDescription = `Continuation of trial ${trialId}, experiment` +
          ` ${experiment.id} (${rawConfig.description})`;
          const newSearcher = {
            max_steps: 100, // TODO add form
            metric: rawConfig.searcher.metric,
            name: rawConfig.searcher.name,
            smaller_is_better: rawConfig.searcher.smaller_is_better,
            source_trial_id: trialId,
          };

          const newHyperparameters = trialHParamsToExperimentHParams(hparams);
```

AFAIK these are the only fields we need to update. @katport  does this sound, right?

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: webui/react/src/components/TrialActions.tsx [hunk 1]
```diff
@@ -10,20 +10,23 @@ import { terminalRunStates } from 'utils/types';
...
   Logs = 'Logs',
   Tensorboard = 'Tensorboard',
 }
 
+interface Props {
+  trial: TrialDetails;
+  onSettled: () => void; // A callback to trigger after an action is done.
+  onClick: {
+    [key in Action]?: () => void;
+  }
```
Comment (hkang1) [inline webui/react/src/components/TrialActions.tsx:24]: **suggestion:** this is an unusual pattern and bit hard to follow. I recommend just the standard:
```
onClick: () => Action;
```
and let the parent handle what to do with the different `Action`s it receives.

File: webui/react/src/types.ts [hunk 2]
```diff
@@ -257,13 +264,15 @@ export interface Step {
   state: RunState;
 }
 
+// FIXME this vs TrialItem
```
Comment (hkang1) [inline webui/react/src/types.ts:267]: I think we discussed this being different right? They can share a common base possibly of:
```
interface Trial {}
interface TrialItem extends Trial {}
interface TrialDetail extends Trial {}
```

File: webui/react/src/components/CreateExperimentModal.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+import { routeAll } from 'routes';
+import { forkExperiment } from 'services/api';
+
+import css from './CreateExperimentModal.module.scss';
+
+export interface CreateExpModalState {
+  visible: boolean;
+  config: string;
+}
```
Comment (hkang1) [inline webui/react/src/components/CreateExperimentModal.tsx:14]: **suggestion:** these two states don't really have anything in common, and by that I mean that when you call `setState`, you are only really changing one or the other and not the two simultaneously. Rather than bundling them into a separate interface, it'll be easier to handle code that treats them separately.
```
interface Props {
  ...
  config: string;
  onCancel: () => void;
  onConfigChange: () => string;
  visible?: boolean;
}
```

Comment (hkang1) [reply webui/react/src/components/CreateExperimentModal.tsx:14]: there is another concern of passing in a function that can mutate a state of a different component directly that's not itself instead of a buffered event handler.

File: webui/react/src/components/CreateExperimentModal.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  const editorOnChange = useCallback((newValue: string) => {
+    setState((existingState: CreateExpModalState) => ({ ...existingState, config: newValue }));
+    setConfigError(undefined);
+  }, [ setState, setConfigError ]);
+
+  const monacoOpts = {
+    minimap: { enabled: false },
+    selectOnLineNumbers: true,
+  };
```
Comment (hkang1) [inline webui/react/src/components/CreateExperimentModal.tsx:37]: **nit:** I've just been putting this directly into the `MonacoEditor` component. Then no need to play connect the dots when looking for it.

File: webui/react/src/pages/TrialDetails.tsx [hunk 5]
```diff
@@ -22,13 +27,62 @@ interface Params {
...
+          const newHyperparameters = trialHParamsToExperimentHParams(hparams);
+
+          rawConfig = {
+            ...rawConfig,
+            description: newDescription,
+            hyperparameters: newHyperparameters,
+            searcher: newSearcher,
+          };
+
+          setContModalState(state => ({ ...state, config: yaml.safeDump(rawConfig) }));
```
Comment (hkang1) [inline webui/react/src/pages/TrialDetails.tsx:70]: **nit:** slight tweak to avoid needing `let`.
```
          const rawConfig = clone(experiment.configRaw);
          const newDescription = `Continuation of trial ${trialId}, experiment` +
          ` ${experiment.id} (${rawConfig.description})`;
          const newSearcher = {
            max_steps: 100, // TODO add form
            metric: rawConfig.searcher.metric,
            name: rawConfig.searcher.name,
            smaller_is_better: rawConfig.searcher.smaller_is_better,
            source_trial_id: trialId,
          };

          const newHyperparameters = trialHParamsToExperimentHParams(hparams);

          setContModalState(state => ({ ...state, config: yaml.safeDump({
            ...rawConfig,
            description: newDescription,
            hyperparameters: newHyperparameters,
            searcher: newSearcher,
          }) }));
```

File: webui/react/src/components/CreateExperimentModal.tsx (hunk 4 above)
Comment (hkang1) [reply webui/react/src/components/CreateExperimentModal.tsx:37]: **chore:** add `scrollBeyondLastLine: false,` to prevent it from scrolling to show a bunch of white space below the last line.

File: webui/react/src/components/CreateExperimentModal.tsx (hunk 3 above)
Comment (hamidzr) [reply webui/react/src/components/CreateExperimentModal.tsx:14]: I had it the other way first but here I'm thinking we could separate writeable states of the component into its own group as I'm anticipating we add more when we add the form on top or more features.

update: I separated the read part of the states into separate props but as we talked I left the set part of it.

Comment (hamidzr) [reply webui/react/src/components/CreateExperimentModal.tsx:14]: I left out the part that creates another handler on top of the setState to let the interface of the setState which allows both passing the new state and a function to set the state but also lower changes to experiment details page as you already have another or open there.

File: webui/react/src/types.ts (hunk 2 above)
Comment (hamidzr) [reply webui/react/src/types.ts]: I went with TrialBase to separate with the other items which don't have the Item suffix.

File: webui/react/src/components/CreateExperimentModal.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  config: string;
+}
+
+interface Props {
+  title: string;
+  okText: string;
+  parentId: number;
+  visible: boolean;
+  config: string;
+  setState: (arg0: SetStateAction<CreateExpModalState>) => void;
```
Comment (hkang1) [inline webui/react/src/components/CreateExperimentModal.tsx:22]: **issue:** TL;DR - avoid passing `setState` to child component and instead pass in a `setState`

The crux problem of this anti-pattern is that the state writing operation is in two places, the parent and the child component. The `on[Action]` pattern is used to keep the state and setState contained in the parent and the child modal will just send events back to the parent for it to handle the source of truth (state). It's also perfectly fine to have multiple `on[Action]` handler because the idea there is to orthogonalize unrelated events from each other, so the handlers do not have to worry about one event from another. This is especially important with the `useCallback` hook, because of the dependency list.

File: webui/react/src/pages/TrialDetails.tsx [hunk 7]
```diff
@@ -24,11 +29,62 @@ const TrialDetailsComp: React.FC = () => {
   const trialId = parseInt(trialIdParam);
   const [ trial, triggerTrialRequest ] =
     useRestApi<TrialDetailsParams, TrialDetails>(getTrialDetails, { id: trialId });
+  const [ ContModalState, setContModalState ] = useState<CreateExpModalState>(
+    { config: 'Loading', visible: false },
+  );
+
```
Comment (hkang1) [inline webui/react/src/pages/TrialDetails.tsx:35]: **suggestion:** separate these into two different `useState`. They are related in that they are both fed into the modal but that's about it. Changes in visibility and changes in config shouldn't alter each other.

Comment (hamidzr) [reply webui/react/src/pages/TrialDetails.tsx:35]: I see that, I can separate them. let's chat about it. Wouldn't react be smart enough to not cause rerenders if only one of these changes? On the other hand, I remember adding each individual hook adds an extra rerender.

**Comment by hamidzr** [comment]
@hkang1  how does https://github.com/determined-ai/determined/pull/971/commits/132ed481e3c05402347139917bf837cdb8a429a4 look?

File: webui/react/src/components/CreateExperimentModal.tsx [hunk 8]
```diff
@@ -1,43 +1,39 @@
...
-
 interface Props {
   title: string;
   okText: string;
   parentId: number;
   visible: boolean;
   config: string;
-  setState: (arg0: SetStateAction<CreateExpModalState>) => void;
+  onVisibleChange: (arg0: boolean) => void;
+  onConfigChange: (arg0: string) => void;
```
Comment (hkang1) [inline webui/react/src/components/CreateExperimentModal.tsx:18]: **super nit:** `visible` and `config` as alternative to `arg0`?
